Cognitive Account - Runtime error: Plugin crashed!

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

<!--- Please keep this note for the community --->

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

<!--- Thank you for keeping this note for the community --->


### Terraform Version

1.8.2

### AzureRM Provider Version

4.14.0

### Affected Resource(s)/Data Source(s)

azurerm_cognitive_account

### Terraform Configuration Files

```hcl
Config (sensitiv data is overwritten with "example"):

{
  "module": "example",
  "mode": "managed",
  "type": "azurerm_cognitive_account",
  "name": "example",
  "provider": "provider[\"registry.terraform.io/hashicorp/azurerm\"]",
  "instances": [
    {
      "schema_version": 0,
      "attributes": {
        "custom_question_answering_search_service_id": null,
        "custom_question_answering_search_service_key": null,
        "custom_subdomain_name": "example",
        "customer_managed_key": [
          {
            "identity_client_id": "",
            "key_vault_key_id": ""
          }
        ],
        "dynamic_throttling_enabled": false,
        "endpoint": "https://example.openai.azure.com/",
        "fqdns": [],
        "id": "example",
        "identity": [
          {
            "identity_ids": [],
            "principal_id": "example",
            "tenant_id": "example",
            "type": "SystemAssigned"
          }
        ],
        "kind": "OpenAI",
        "local_auth_enabled": true,
        "location": "example",
        "metrics_advisor_aad_client_id": null,
        "metrics_advisor_aad_tenant_id": null,
        "metrics_advisor_super_user_name": null,
        "metrics_advisor_website_name": null,
        "name": "example",
        "network_acls": [
          {
            "default_action": "Deny",
            "ip_rules": [
              "example"
            ],
            "virtual_network_rules": [
              {
                "ignore_missing_vnet_service_endpoint": false,
                "subnet_id": "example"
              },
              {
                "ignore_missing_vnet_service_endpoint": false,
                "subnet_id": "example"
              }
            ]
          }
        ],
        "outbound_network_access_restricted": false,
        "primary_access_key": "example",
        "public_network_access_enabled": true,
        "qna_runtime_endpoint": null,
        "resource_group_name": "example",
        "secondary_access_key": "example",
        "sku_name": "S0",
        "storage": [],
        "tags": {},
        "timeouts": null
      },
      "sensitive_attributes": [
        [
          {
            "type": "get_attr",
            "value": "secondary_access_key"
          }
        ],
        [
          {
            "type": "get_attr",
            "value": "custom_question_answering_search_service_key"
          }
        ],
        [
          {
            "type": "get_attr",
            "value": "primary_access_key"
          }
        ]
      ],
      "dependencies": [
        "example"
      ]
    }
  ]
}

Following message during the tf apply step:
╷
│ Error: Plugin did not respond
│ 
│   with module.example.azurerm_cognitive_account.example,
│   on ../../../modules/example/module.tf line 18, in resource "azurerm_cognitive_account" "example":
│   18: resource "azurerm_cognitive_account" "example" {
│ 
│ The plugin encountered an error, and failed to respond to the
│ plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.
╵
```


### Debug Output/Panic Output

```shell
Stack trace from the terraform-provider-azurerm_v4.14.0_x5 plugin:

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x54dfdba]

goroutine 85 [running]:
github.com/hashicorp/terraform-provider-azurerm/internal/services/cognitive.expandCognitiveAccountCustomerManagedKey({0xc00316d370?, 0x82c8b7e?, 0x17?})
	github.com/hashicorp/terraform-provider-azurerm/internal/services/cognitive/cognitive_account_resource.go:847 +0x15a
github.com/hashicorp/terraform-provider-azurerm/internal/services/cognitive.resourceCognitiveAccountUpdate(0xc0030adc80, {0x71ed100, 0xc001f23208})
	github.com/hashicorp/terraform-provider-azurerm/internal/services/cognitive/cognitive_account_resource.go:471 +0xc12
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).update(0x8ab4548?, {0x8ab4548?, 0xc003106990?}, 0xd?, {0x71ed100?, 0xc001f23208?})
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.35.0/helper/schema/resource.go:828 +0x15f
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc000f31ce0, {0x8ab4548, 0xc003106990}, 0xc001b885b0, 0xc0030adb00, {0x71ed100, 0xc001f23208})
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.35.0/helper/schema/resource.go:947 +0x83a
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc0009d4a20, {0x8ab4548?, 0xc0031068d0?}, 0xc001f24fa0)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.35.0/helper/schema/grpc_provider.go:1155 +0xd5c
github.com/hashicorp/terraform-plugin-mux/tf5muxserver.(*muxServer).ApplyResourceChange(0xc0001baf80, {0x8ab4548?, 0xc003106600?}, 0xc001f24fa0)
	github.com/hashicorp/terraform-plugin-mux@v0.17.0/tf5muxserver/mux_server_ApplyResourceChange.go:36 +0x193
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0xc000aca960, {0x8ab4548?, 0xc0030edbc0?}, 0xc001b6fce0)
	github.com/hashicorp/terraform-plugin-go@v0.25.0/tfprotov5/tf5server/server.go:865 +0x3d0
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0x7f9c8a0, 0xc000aca960}, {0x8ab4548, 0xc0030edbc0}, 0xc0030acf00, 0x0)
	github.com/hashicorp/terraform-plugin-go@v0.25.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:611 +0x1a6
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0001b6e00, {0x8ab4548, 0xc0030edb30}, {0x8ae29c0, 0xc001c65520}, 0xc0030f58c0, 0xc001b90750, 0xe1b8278, 0x0)
	google.golang.org/grpc@v1.67.1/server.go:1394 +0xe49
google.golang.org/grpc.(*Server).handleStream(0xc0001b6e00, {0x8ae29c0, 0xc001c65520}, 0xc0030f58c0)
	google.golang.org/grpc@v1.67.1/server.go:1805 +0xe8b
google.golang.org/grpc.(*Server).serveStreams.func2.1()
	google.golang.org/grpc@v1.67.1/server.go:1029 +0x8b
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 29
	google.golang.org/grpc@v1.67.1/server.go:1040 +0x125

Error: The terraform-provider-azurerm_v4.14.0_x5 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```


### Expected Behaviour

We have one code source and we deploy multiple instances of the azurerm_cognitive_account. The other instances are working fine, just the one we tried out the customer managed key encryption experience crashes. We already switched back to Microsoft managed key encryption. 

Since this switch, on this instance the Microsoft ARM Template has following entry: 
                "encryption": {
                    "keySource": "Microsoft.CognitiveServices" 

The other instances do not have such a entry.
I guess the specific go file in the services section can not handle this type of information.

### Actual Behaviour

Get a crash!

### Steps to Reproduce

terraform apply

### Important Factoids

_No response_

### References

A similar issue in the backup vault resource which is already solved: https://github.com/hashicorp/terraform-provider-azurerm/issues/27751

--- a/internal/services/cognitive/cognitive_account_resource.go
+++ b/internal/services/cognitive/cognitive_account_resource.go
@@ -9,6 +9,7 @@ import (
 	"log"
 	"time"
 
+	"github.com/hashicorp/go-azure-helpers/lang/pointer"
 	"github.com/hashicorp/go-azure-helpers/lang/response"
 	"github.com/hashicorp/go-azure-helpers/resourcemanager/commonids"
 	"github.com/hashicorp/go-azure-helpers/resourcemanager/commonschema"
@@ -829,7 +830,9 @@ func flattenCognitiveAccountStorage(input *[]cognitiveservicesaccounts.UserOwned
 
 func expandCognitiveAccountCustomerManagedKey(input []interface{}) *cognitiveservicesaccounts.Encryption {
 	if len(input) == 0 || input[0] == nil {
-		return nil
+		return &cognitiveservicesaccounts.Encryption{
+			KeySource: pointer.To(cognitiveservicesaccounts.KeySourceMicrosoftPointCognitiveServices),
+		}
 	}
 
 	v := input[0].(map[string]interface{})
@@ -857,7 +860,7 @@ func expandCognitiveAccountCustomerManagedKey(input []interface{}) *cognitiveser
 }
 
 func flattenCognitiveAccountCustomerManagedKey(input *cognitiveservicesaccounts.Encryption) ([]interface{}, error) {
-	if input == nil {
+	if input == nil || pointer.From(input.KeySource) == cognitiveservicesaccounts.KeySourceMicrosoftPointCognitiveServices {
 		return []interface{}{}, nil
 	}
 

--- a/internal/services/cognitive/cognitive_account_resource.go
+++ b/internal/services/cognitive/cognitive_account_resource.go
@@ -473,6 +473,17 @@ func resourceCognitiveAccountUpdate(d *pluginsdk.ResourceData, meta interface{})
 		},
 		Tags: tags.Expand(d.Get("tags").(map[string]interface{})),
 	}
+
+	if d.HasChanges("customer_managed_key") {
+		old, new := d.GetChange("customer_managed_key")
+		// Remove `customer_managed_key` (switch using a customer managed key to microsoft managed), and explicitly specify KeySource as `Microsoft.CognitiveServices`.
+		if len(old.([]interface{})) > 0 && len(new.([]interface{})) == 0 {
+			props.Properties.Encryption = &cognitiveservicesaccounts.Encryption{
+				KeySource: pointer.To(cognitiveservicesaccounts.KeySourceMicrosoftPointCognitiveServices),
+			}
+		}
+	}
+
 	identityRaw := d.Get("identity").([]interface{})
 	identity, err := identity.ExpandSystemAndUserAssignedMap(identityRaw)
 	if err != nil {
@@ -830,9 +841,7 @@ func flattenCognitiveAccountStorage(input *[]cognitiveservicesaccounts.UserOwned
 
 func expandCognitiveAccountCustomerManagedKey(input []interface{}) *cognitiveservicesaccounts.Encryption {
 	if len(input) == 0 || input[0] == nil {
-		return &cognitiveservicesaccounts.Encryption{
-			KeySource: pointer.To(cognitiveservicesaccounts.KeySourceMicrosoftPointCognitiveServices),
-		}
+		return nil
 	}
 
 	v := input[0].(map[string]interface{})

--- a/internal/services/cognitive/cognitive_account_resource.go
+++ b/internal/services/cognitive/cognitive_account_resource.go
@@ -834,6 +834,10 @@ func expandCognitiveAccountCustomerManagedKey(input []interface{}) *cognitiveser
 
 	v := input[0].(map[string]interface{})
 	keyId, _ := keyVaultParse.ParseOptionallyVersionedNestedItemID(v["key_vault_key_id"].(string))
+	if keyId == nil {
+		return nil
+	}
+
 	keySource := cognitiveservicesaccounts.KeySourceMicrosoftPointKeyVault
 
 	var identity string

--- a/internal/services/cognitive/cognitive_account_resource_test.go
+++ b/internal/services/cognitive/cognitive_account_resource_test.go
@@ -382,6 +382,13 @@ func TestAccCognitiveAccount_customerManagedKey(t *testing.T) {
 			),
 		},
 		data.ImportStep(),
+		{
+			Config: r.customerManagedKeyUpdate(data),
+			Check: acceptance.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+			),
+		},
+		data.ImportStep(),
 	})
 }
 
@@ -1101,6 +1108,87 @@ resource "azurerm_cognitive_account" "test" {
   customer_managed_key {
     key_vault_key_id   = azurerm_key_vault_key.test.id
     identity_client_id = azurerm_user_assigned_identity.test.client_id
+  }
+}
+`, data.RandomInteger, data.Locations.Secondary, data.RandomString, data.RandomString, data.RandomString, data.RandomInteger, data.RandomInteger)
+}
+
+func (CognitiveAccountResource) customerManagedKeyUpdate(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+provider "azurerm" {
+  features {
+    key_vault {
+      purge_soft_delete_on_destroy       = false
+      purge_soft_deleted_keys_on_destroy = false
+    }
+  }
+}
+
+data "azurerm_client_config" "current" {}
+
+resource "azurerm_resource_group" "test" {
+  name     = "acctestRG-cognitive-%d"
+  location = "%s"
+}
+
+resource "azurerm_user_assigned_identity" "test" {
+  resource_group_name = azurerm_resource_group.test.name
+  location            = azurerm_resource_group.test.location
+  name                = "%s"
+}
+
+resource "azurerm_key_vault" "test" {
+  name                     = "acctestkv%s"
+  location                 = azurerm_resource_group.test.location
+  resource_group_name      = azurerm_resource_group.test.name
+  tenant_id                = data.azurerm_client_config.current.tenant_id
+  sku_name                 = "standard"
+  purge_protection_enabled = true
+
+  access_policy {
+    tenant_id = data.azurerm_client_config.current.tenant_id
+    object_id = data.azurerm_client_config.current.object_id
+    key_permissions = [
+      "Get", "Create", "Delete", "List", "Restore", "Recover", "UnwrapKey", "WrapKey", "Purge", "Encrypt", "Decrypt", "Sign", "Verify", "GetRotationPolicy"
+    ]
+    secret_permissions = [
+      "Get",
+    ]
+  }
+
+  access_policy {
+    tenant_id = azurerm_user_assigned_identity.test.tenant_id
+    object_id = azurerm_user_assigned_identity.test.principal_id
+    key_permissions = [
+      "Get", "Create", "Delete", "List", "Restore", "Recover", "UnwrapKey", "WrapKey", "Purge", "Encrypt", "Decrypt", "Sign", "Verify", "GetRotationPolicy"
+    ]
+    secret_permissions = [
+      "Get",
+    ]
+  }
+}
+
+resource "azurerm_key_vault_key" "test" {
+  name         = "acctestkvkey%s"
+  key_vault_id = azurerm_key_vault.test.id
+  key_type     = "RSA"
+  key_size     = 2048
+  key_opts     = ["decrypt", "encrypt", "sign", "unwrapKey", "verify", "wrapKey"]
+}
+
+resource "azurerm_cognitive_account" "test" {
+  name                  = "acctest-cogacc-%d"
+  location              = azurerm_resource_group.test.location
+  resource_group_name   = azurerm_resource_group.test.name
+  kind                  = "SpeechServices"
+  sku_name              = "S0"
+  custom_subdomain_name = "acctest-cogacc-%d"
+
+  identity {
+    type = "SystemAssigned, UserAssigned"
+    identity_ids = [
+      azurerm_user_assigned_identity.test.id
+    ]
   }
 }
 `, data.RandomInteger, data.Locations.Secondary, data.RandomString, data.RandomString, data.RandomString, data.RandomInteger, data.RandomInteger)

